hack/update-unwanted-dependencies.sh

```diff
@@ -0,0 +1,91 @@
+
+# This script checks unwanted dependencies. It checks if any items in the
+# "unwanted but still present" list(.unwanted_dependencies) are no longer present 
+# and prompts to move them to the "absent and not allowed" list(.forbidden_dependencies)
+# It also checks if any items in the "absent and not allowed" list are present and fails.
+# Usage: `hack/update-unwanted-dependencies.sh`.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+# Explicitly opt into go modules, even though we're inside a GOPATH directory
+export GO111MODULE=on
+# Explicitly set GOFLAGS to ignore vendor, since GOFLAGS=-mod=vendor breaks dependency resolution while rebuilding vendor
+export GOFLAGS=-mod=mod
+# Detect problematic GOPROXY settings that prevent lookup of dependencies
+if [[ "${GOPROXY:-}" == "off" ]]; then
+  kube::log::error "Cannot run with \$GOPROXY=off"
+  exit 1
+fi
+
+# We store unwanted dependencies which is still present in .unwanted_dependencies
+unwanted_dependencies_file="${KUBE_ROOT}/hack/.unwanted_dependencies"
+
+# We store not-allowed dependencies in .forbidden_dependencies
+not_allowed_dependencies_file="${KUBE_ROOT}/hack/.forbidden_dependencies"
+
+not_allowed_dependencies=()
+while IFS='' read -r line; do
+  not_allowed_dependencies+=("$line")
+done < <(cat "${not_allowed_dependencies_file}")
+
+# check unwanted dependencies, and update the .forbidden_dependencies and .unwanted_dependencies file if some unwanted dependencies are removed.
+unwanted_dependencies=()
+while IFS='' read -r line; do
+  unwanted_dependencies+=("$line")
+done < <(cat "${unwanted_dependencies_file}")
+
+# a flag to know if need updated
+changed=false
+
+new_unwanted_dependencies=()
+for unwanted_repo in "${unwanted_dependencies[@]}"; do
+  deps_on_unwanted_repo=$(go mod graph | grep " ${unwanted_repo}@" || echo "")
+  if [ -n "${deps_on_unwanted_repo}" ]; then
+  	# still exist
+  	new_unwanted_dependencies+=("${unwanted_repo}")
+  else
+  	changed=true
+	  not_allowed_dependencies+=("${unwanted_repo}")
+    echo "Move ${unwanted_repo} to not_allowed_dependencies"
+    echo ""
+  fi
+done
+
+if [ "${changed}" = true ]; then
+  echo "Update unwanted and not allowed dependencies"
+
+  echo -n "" > "${unwanted_dependencies_file}.sorted"
+  for unwanted_repo in "${new_unwanted_dependencies[@]}"; do
+  	echo "${unwanted_repo}" >> "${unwanted_dependencies_file}.sorted"
+  done
+
+  echo -n "" > "${not_allowed_dependencies_file}.sorted"
+  for forbidden_repo in "${not_allowed_dependencies[@]}"; do
+  	echo "${forbidden_repo}" >> "${not_allowed_dependencies_file}.sorted"
+  done
+
+  sort "${unwanted_dependencies_file}.sorted" | uniq > "${unwanted_dependencies_file}"
+  sort "${not_allowed_dependencies_file}.sorted" | uniq > "${not_allowed_dependencies_file}"
+  rm "${unwanted_dependencies_file}.sorted" 
+  rm "${not_allowed_dependencies_file}.sorted" 
+fi
```

Comment:
I think this whole section can be simplified with a single loop, no intermediate `.sorted` files and no `changed` variable:

```bash
unwanted_dependencies=()
while IFS='' read -r unwanted_repo; do
  deps_on_unwanted_repo=$(go mod graph | grep " ${unwanted_repo}@" || echo "")
  if [ -n "${deps_on_unwanted_repo}" ]; then
    unwanted_dependencies+=("${unwanted_repo}")
  else
    forbidden_dependencies+=("${unwanted_repo}")
    echo "Moved ${unwanted_repo} to .forbidden_dependencies"
  fi
done < <(cat "${unwanted_dependencies_file}")

echo "${unwanted_dependencies[*]}" | tr " " "\n" | sort | uniq > "${unwanted_dependencies_file}"
echo "${forbidden_dependencies[*]}" | tr " " "\n" | sort | uniq > "${not_allowed_dependencies_file}"
```

Reply:
Good point.